### PR TITLE
Add complex type

### PR DESCRIPTION
**Pull request by sbinet**

### CODE REVIEW

File: eval/op.go [hunk 1]
```diff
@@ -242,6 +258,10 @@ func binOp(op token.Token, x, y interface{}) (interface{}, error) {
 			case UntypedInt:
 				z := big.NewInt(0)
 				return UntypedInt{z.Sub(x.Int, y.Int)}, nil
+			case UntypedComplex:
+				re := big.NewFloat(0)
+				xf := big.NewFloat(float64(x.Int.Int64()))
+				return UntypedComplex{re.Sub(xf, y.Real), big.NewFloat(0)}, nil
```
Comment (crawshaw) [inline eval/op.go]: Why are you ignoring the Imag component of y?

File: eval/eval.go [hunk 2]
```diff
@@ -150,6 +150,90 @@ func New(path string) *Program {
 	})
 	addUniverse("make", p.builtinMake)
 	addUniverse("new", p.builtinNew)
+	addUniverse("complex", func(re, im interface{}) interface{} {
```
Comment (crawshaw) [inline eval/eval.go]: As there's a bit of code here, how about defining it in a method, `builtinComplex`?

File: eval/testdata/complex.ng [hunk 3]
```diff
@@ -1,5 +1,21 @@
...
+complex(1.1, 2.2)
+complex(1.1, float32(2))
+complex(1.1, float64(2))
+
+complex(float32(2), float32(3))
+
+complex(float64(2), float64(3))
+
 // TODO: c := complex128(0) + complex128(complex64(3.5))
 // TODO: real(1 + 2i) == 1.0
```
Comment (crawshaw) [inline eval/testdata/complex.ng]: Looks like these will work now. Want to remove the comments and replace the print at the bottom with something like:

if (some addition of complex numbers) == (some converted value) {
    print("OK")
}

File: format/expr.go [hunk 4]
```diff
@@ -22,6 +22,45 @@ type printer struct {
...
 func (p *printer) expr(e expr.Expr) {
 	switch e := e.(type) {
+	case *expr.Binary:
+		WriteExpr(p.buf, e.Left)
+		p.buf.WriteString(e.Op.String())
+		WriteExpr(p.buf, e.Right)
+	case *expr.Unary:
+		p.buf.WriteString(e.Op.String())
+		WriteExpr(p.buf, e.Expr)
+	case *expr.Bad:
```
Comment (crawshaw) [inline format/expr.go]: Before these would fall through to the default case and print the data structure. I think that's still useful. How about commenting this block of cases out and putting a TODO on it?

File: eval/eval.go (hunk 2 above)
Comment (sbinet) [reply eval/eval.go]: done.

File: eval/op.go (hunk 1 above)
Comment (sbinet) [reply eval/op.go]: copy-pasta...
done.

File: eval/testdata/complex.ng (hunk 3 above)
Comment (sbinet) [reply eval/testdata/complex.ng]: unfortunately, no.

```
ng> complex(1,2)
complex128((1+2i))

ng> c := complex(1,2)
ng eval panic: TODO rtype for Complex
goroutine 1 [running]:
runtime/debug.Stack(0x26, 0x0, 0x0)
	/usr/lib/go/src/runtime/debug/stack.go:24 +0xa7
runtime/debug.PrintStack()
	/usr/lib/go/src/runtime/debug/stack.go:16 +0x22
neugram.io/ng/eval.(*Program).Eval.func1(0xc420156090, 0xc42010b718, 0xc42010b700)
	/home/binet/work/gonum/src/neugram.io/ng/eval/eval.go:411 +0x204
panic(0x8dfa00, 0xa138c0)
	/usr/lib/go/src/runtime/panic.go:491 +0x283
neugram.io/ng/eval.(*reflector).ToRType(0xc420148040, 0xd81d00, 0xa13cf0, 0xc4200f2240, 0x1)
	/home/binet/work/gonum/src/neugram.io/ng/eval/eval.go:1423 +0x265c
neugram.io/ng/eval.(*Program).evalStmt(0xc420156090, 0xd80cc0, 0xc4200f4780, 0x0, 0x0, 0x0)
	/home/binet/work/gonum/src/neugram.io/ng/eval/eval.go:466 +0x18cf
neugram.io/ng/eval.(*Program).Eval(0xc420156090, 0xd80cc0, 0xc4200f4780, 0xc42005e5a0, 0x0, 0x0, 0x0, 0xd80380, 0xc420148200)
	/home/binet/work/gonum/src/neugram.io/ng/eval/eval.go:424 +0x1fc
main.handleResult(0x1, 0xc4201481b0, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/binet/work/gonum/src/neugram.io/ng/ng.go:338 +0x734
main.loop()
	/home/binet/work/gonum/src/neugram.io/ng/ng.go:331 +0x602
main.main()
	/home/binet/work/gonum/src/neugram.io/ng/ng.go:119 +0x35c
ng: ng eval panic: TODO rtype for Complex

ng> 
```

File: format/expr.go (hunk 4 above)
Comment (sbinet) [reply format/expr.go]: done.

**Comment by crawshaw** [comment]
LGTM.

I'm not sure what the normal process is now in github reviews. I'll go ahead and "rebase and merge" this pull request. I would be tempted to squash the review changes commits a-la gerrit, but I don't want to squash all of your commits into one as you have them broken up well.

**Comment by crawshaw** [comment]
Also I guess travis doesn't run on pull requests? Well we can see the effect when I merge. :-)

**Comment by sbinet** [comment]
yeah... github is catching up but there are still a few things to implement to match gerrit :)

I wouldn't have been upset to be resquested to squash a few commits together :)
(say, `format: expr` and `format: add ...`)

**Comment by shurcooL** [comment]
> Also I guess travis doesn't run on pull requests?

It normally does. Perhaps you have it turned off somehow? According to https://docs.travis-ci.com/user/pull-requests/:

> you can disable Build on Pull Requests from your repository settings.

You can check your settings at https://travis-ci.org/neugram/ng/settings, it should look something like this:

image

**Comment by crawshaw** [comment]
It's turned on. The "Pull requests" tab claims there are no outstanding requests even though @sbinet has one open right now. Maybe a travis bug?

**Comment by sbinet** [comment]
Yeah. I think Travis doesn't detect pull requests that come from a fork hosted by an organization (sbinet-gonum is an organization I created for my gonum related work)

I haven't figured out yet wherever that's really the root of the issue... But that's my current hunch.

**Comment by shurcooL** [comment]
@sbinet I wanted to test that theory, so I made a PR from a non-organization account in #25. It seems confirms it, because my PR got picked up by Travis without issues.

**Comment by sbinet** [comment]
I have also tested it (creating a dumb "orig-org/repo", and a "fork-org/repo", etc...).
turns out, for PR initiated by the "fork-org/repo" to be picked up by travis, one needs to enable travis on the "fork-org/repo" too (via the control panel of travis-ci.)

case closed :)

**Pull request merged**